server/entity/arrow.go

```diff
@@ -0,0 +1,333 @@
+
+	c *ProjectileComputer
+}
+
+// NewArrow ...
+func NewArrow(pos mgl64.Vec3, yaw, pitch float64, owner world.Entity, critical, disallowPickup, obtainArrowOnPickup bool, tip potion.Potion) *Arrow {
+	a := &Arrow{
+		yaw:   yaw,
+		pitch: pitch,
+		c: &ProjectileComputer{&MovementComputer{
+			Gravity:           0.05,
+			Drag:              0.01,
+			DragBeforeGravity: true,
+		}},
+		baseDamage:          2.0,
+		disallowPickup:      disallowPickup,
+		obtainArrowOnPickup: obtainArrowOnPickup,
+		critical:            critical,
+		owner:               owner,
+		tip:                 tip,
+	}
+	a.transform = newTransform(a, pos)
+	return a
+}
+
+// Name ...
+func (a *Arrow) Name() string {
+	return "Arrow"
+}
+
+// EncodeEntity ...
+func (a *Arrow) EncodeEntity() string {
+	return "minecraft:arrow"
+}
+
+// CollisionPos returns the position of the block the arrow collided with. If the arrow has not collided with any
+// blocks, it returns false for it's second parameter.
+func (a *Arrow) CollisionPos() (cube.Pos, bool) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	return a.collidedBlockPos, a.collidedBlock != nil
+}
+
+// Critical returns the critical state of the arrow, which can result in more damage and extra particles while in air.
+func (a *Arrow) Critical() bool {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	return a.critical
+}
+
+// SetCritical sets the critical state of the arrow to true, which can result in more damage and extra particles while
+// in air.
+func (a *Arrow) SetCritical(critical bool) {
+	a.mu.Lock()
+	a.critical = critical
+	pos := a.pos
+	a.mu.Unlock()
+
+	for _, v := range a.World().Viewers(pos) {
+		v.ViewEntityState(a)
+	}
+}
+
+// BaseDamage returns the base damage the arrow will deal, before accounting for velocity.
+func (a *Arrow) BaseDamage() float64 {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	return a.baseDamage
+}
+
+// SetBaseDamage sets the base damage the arrow will deal, before accounting for velocity.
+func (a *Arrow) SetBaseDamage(baseDamage float64) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	a.baseDamage = baseDamage
+}
+
+// ShouldDisallowPickup returns true if the arrow should not be picked up by players.
+func (a *Arrow) ShouldDisallowPickup() bool {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	return a.disallowPickup
+}
+
+// DisallowPickup sets whether the arrow should or should not be picked up by players.
+func (a *Arrow) DisallowPickup(disallowPickup bool) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	a.disallowPickup = disallowPickup
+}
+
+// ShouldObtainArrowOnPickup returns true if the arrow should be obtained when picked up by players.
+func (a *Arrow) ShouldObtainArrowOnPickup() bool {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	return a.obtainArrowOnPickup
+}
+
+// ObtainArrowOnPickup sets whether the arrow should or should not be obtained when picked up by players.
+func (a *Arrow) ObtainArrowOnPickup(obtainArrowOnPickup bool) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	a.obtainArrowOnPickup = obtainArrowOnPickup
+}
```

Comment:
Lots of methods here that I don't think we need necessarily. Is there any merit of having these getters and especially setters? These are passed in the `NewArrow` method and it sounds like a better idea to have all these things, like whether an arrow is received on pickup and if pickup is allowed, be immutable.

Reply:
Changed. Will keep a way to set and get base damage though, because I think that could be useful to some people.